File: 0133-clone-graph/0133-clone-graph.py

class Solution:
    def clone(self, node, clones={}):
        if not node:
            return None

        # If node is already cloned, return the clone
        if node in clones:
            return clones[node]

        # It is important to make the node of the graph as it matters further for making the neighbors
        # Create a new clone of the current node
        clone_node = Node(node.val)
        clones[node] = clone_node

        # Clone the neighbors recursively
        for neighbor in node.neighbors:
            clone_node.neighbors.append(self.clone(neighbor, clones))

        return clone_node

    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        return self.clone(node)

File: 0133-clone-graph/0133-clone-graph.js
/**
 * // Definition for a _Node.
 * function _Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */

const clone = (node, clones = {}) => {
    if(!node){
        return null
    }

    if(clones[node.val]){
        return clones[node.val]
    }

    const cloneNode = new _Node(node.val)
    clones[node.val] = cloneNode

    for (let neighbor of node.neighbors) {
        cloneNode.neighbors.push(clone(neighbor, clones))
    }

    return cloneNode
}

/**
 * @param {_Node} node
 * @return {_Node}
 */
var cloneGraph = function(node) {
    return clone(node)
};